File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    void expand(string& s, int l, int r, int& start, int& end) {
        while(l >= 0 && r < s.size() && s[l] == s[r]) {
            l--;
            r++;
        }

        l++;
        r--;

        if(r - l > end - start) {
            start = l;
            end = r;
        }
    }

    string longestPalindrome(string s) {
        int n = s.size();
        int start = 0, end = 0;

        for(int i=0;i<n;i++) {
            expand(s, i, i, start, end);
            expand(s, i, i+1, start, end);
        }

        return s.substr(start, end - start + 1);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public int expand(String s, int l, int r) {
        while(l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }

        return r - l - 1;
    }
    
    public String longestPalindrome(String s) {
        int n = s.length();
        int start = 0, end = 0;

        for(int i=0;i<n;i++) {
            int oddLength = expand(s, i, i);
            int evenLength = expand(s, i, i+1);
            int maxi = Math.max(oddLength, evenLength);

            if(maxi > (end - start + 1)) {
                start = i - (maxi - 1) / 2;
                end = i + maxi / 2;
            }
        }

        return s.substring(start, end + 1);
    }
}